Update target link in core/image when an image is selected

SeanDS opened the pull request:
Closes #13749.

## Description
The image block currently assumes that the default media target is always `none`, so it doesn't update the target link when the user selects an image; it only does this when the user explicitly changes the target using the drop-down box in the sidebar (using the `onSetLinkDestination` callback). This means plugins cannot control the default link target by simply setting `linkDestination.default` to something other than `none`: the drop-down box will show the new default, but the target link will not be properly set when an image is chosen/uploaded (it will still be blank, and the saved post will not therefore link to e.g. the media or attachment page).

This patch simply runs the callback `onSetLinkDestination`, which is already called when a new target selection is made by the user explicitly, when an image is uploaded or selected from the media manager. This ensures that if a plugin changes the default link target, the URL for that target is properly set when the image is selected.

## How has this been tested?
I used the latest development release of Gutenberg (cloned at a59aaf24ce0e081e772e19a2ce0e5c88afa62c77), and WordPress 5.0.3, using VVV. I created a plugin to change the default link target to `media`:

```javascript
function modifyImageLinkDestinationDefault( settings, name ) {
    // Link target to override core default.
    var target = "media";

    if ( name == "core/image" ) {
        // Set image block default destination.
        settings.attributes.linkDestination.default = target;
    } else if ( name == "core/gallery" ) {
        // Set gallery block default destination.
        settings.attributes.linkTo.default = target;
    }

    return settings;
}

// Add settings filter to block registration.
wp.hooks.addFilter(
    "blocks.registerBlockType",
    "my-plugin/modify-image-link-destination-default",
    modifyImageLinkDestinationDefault
);
```

and enqueued it on `enqueue_block_editor_assets` using:

```php
wp_enqueue_script(
	'modify-image-link-destination-default',
	'/path/to/plugin/js/index.js',
	array(
		'wp-hooks',
	)
);
```

## Screenshots
Default behaviour. My plugin changes the default target, but the link URL is not properly set when I select the image:
![image](https://user-images.githubusercontent.com/5225190/52522385-a4905500-2c7c-11e9-819e-b78888ec110e.png)

Patch applied, the link URL is properly set when an image is selected:
![image](https://user-images.githubusercontent.com/5225190/52522343-056b5d80-2c7c-11e9-821a-b211f878449c.png)

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate.

SeanDS committed:
Update target link in core/image when an image is selected

See #13749.

talldan commented on packages/block-library/src/image/edit.js:
```diff
@@ -188,6 +188,9 @@ class ImageEdit extends Component {
 		} );
+
+		// Update target link in case the default has been changed (e.g. by a plugin).
+		this.onSetLinkDestination( this.props.attributes.linkDestination );
```
This results in two consecutive calls to `setAttributes`, which is not great because every call to `setAttributes` creates a new undo state for the user. Ideally all attributes updates would happen in the call to setAttributes on line 186. You'd probably need to extract the contents of `onSetLinkDestination` into a separate function that returns an object that can be merged into the state (the same way that `pickRelevantMediaFiles` works).

SeanDS replied:
Thanks. I've looked into this and it's not as simple as it seems. I split the functions as you said...

```php
	onSelectImage( media ) {
		if ( ! media || ! media.url ) {
			this.props.setAttributes( {
				url: undefined,
				alt: undefined,
				id: undefined,
				caption: undefined,
			} );
			return;
		}

		this.setState( {
			isEditing: false,
		} );

		this.props.setAttributes( {
			...pickRelevantMediaFiles( media ),
			...this.buildLinkAttributes( this.props.attributes.linkDestination ),
			width: undefined,
			height: undefined,
		} );
	}

	onSetLinkDestination( value ) {
		this.props.setAttributes( this.buildLinkAttributes( value ) );
	}

	buildLinkAttributes( destination ) {
		let href;

		if ( destination === LINK_DESTINATION_NONE ) {
			href = undefined;
		} else if ( destination === LINK_DESTINATION_MEDIA ) {
			href = ( this.props.image && this.props.image.source_url ) || this.props.attributes.url;
		} else if ( destination === LINK_DESTINATION_ATTACHMENT ) {
			href = this.props.image && this.props.image.link;
		} else {
			href = this.props.attributes.href;
		}

		return {
			linkDestination: destination,
			href,
		};
	}
```

...but it didn't work - the URL field was empty after selecting an image. I think the problem is that the callbacks bound to the attributes set by `onSelectImage` (using keys from `pickRelevantMediaFiles`) change the values used later by my explicit call to `onSetLinkDestination`. By combining the attribute updates from `pickRelevantMediaFiles` and `onSetLinkDestination` into one `setAttributes` call, like in the above code, the values aren't the same as in the original PR because some haven't yet updated by that part of the execution. Splitting the logic across the two calls to `setAttributes` accidentally worked pretty well to avoid this.

One way to fix this is to explicitly pass the relevant parameters into my new function, e.g. pass `media` in to `buildLinkAttributes`, but this is getting a bit messy. Is there instead some way to set attributes, letting all of the callbacks bound to those attribute fire, before "committing" the attributes later to generate the undo action? Or do you have any other ideas for solving this?

talldan replied:
Ah, that's unfortunate. 

> Is there instead some way to set attributes, letting all of the callbacks bound to those attribute fire, before "committing" the attributes later to generate the undo action?

Kind of like a batch update. As far as I'm aware nothing like that exists.

> One way to fix this is to explicitly pass the relevant parameters into my new function

That seems like the best option, maybe it just accepts the full attributes. Something like this could be the usage:
```
const mediaAttributes = pickRelevantMediaFiles( media );
const linkAttributes = buildLinkAttributes( 
    this.props.attributes.linkDestination,
    { ...this.props.attributes, ...mediaAttributes }
);
this.props.setAttributes( {
    ...mediaAttributes,
    ...linkAttributes,
    width: undefined,
    height: undefined,
} );
```

Usage in `onSetLinkDestination` would still be fairly similar:
```
this.props.setAttributes( this.buildLinkAttributes( value, this.props.attributes ) );
```

talldan reviewed (changes requested):
Hey @SeanDS, thanks for the contribution, I've left a comment about the consecutive calls to `setAttributes`.

SeanDS committed:
Remove second call to setAttributes.

This avoids creating multiple undo operations when the image is uploaded/selected. Because
the media attributes are used to build the URL, the code in onSetLinkDestination has been
broken out into its own function to be used by onSelectImage as well.

SeanDS commented:
@talldan, thanks for the pointers. I've updated the code and tested it and it seems to work.

One thing I wasn't 100% sure about was how I deal with situations when destination is not one of the defined constants (i.e. the `else` block in `buildLinkAttributes`. I've left it as using `this.props.attributes.href` like before. I guess since this property isn't derived from the selected image, this is fine?

SeanDS commented:
Is this good to go? Or any changes needed?

talldan commented:
@SeanDS Sorry for the delay, will give this another review and test.

> One thing I wasn't 100% sure about was how I deal with situations when destination is not one of the defined constants (i.e. the `else` block in `buildLinkAttributes`. I've left it as using `this.props.attributes.href` like before. I guess since this property isn't derived from the selected image, this is fine?

I think the aim there is to leave the href unchanged if the link destination is a custom one or an unknown value.

talldan commented on packages/block-library/src/image/edit.js:
```diff
@@ -183,30 +183,41 @@ class ImageEdit extends Component {
+		} else if ( destination === LINK_DESTINATION_MEDIA ) {
+			href = attributes.source_url || attributes.url;
+		} else if ( destination === LINK_DESTINATION_ATTACHMENT ) {
+			href = attributes.link;
```
The image block doesn't have `link` or `source_url` attributes. It looks like these should be `this.props.image.link` and `this.props.image.source_url`

SeanDS replied:
Hmm, I get an `this.props.image is null` error. I think this is again a case of the information not yet being available to set the link target like before, and it'll probably need another hack to get it right.

I'm starting to think this is a lot more work than it's worth. It's unfortunate that this default behaviour cannot be changed easily, because I believe having an image link to the larger version is a sane default that many users will want (indeed, it's the default in the classic editor, so it's what they'd also probably expect).

I might look into this again at some point but I can't dedicate much more time to fixing this right now. If someone else wants to have a go, please feel free!

jesusangel replied:
I agree with you. I've come to this pull request while I was searching for a way to change the default behaviour of the Gutenberg image block.

talldan reviewed (changes requested):
I spotted an issue in testing - the Link To 'Attachment' option no longer works. Left a comment mentioning where the issue is in the code.

Also in testing I noticed a few things about the general approach of using a plugin to change attribute defaults in this way:
- Existing image blocks (added before the plugin was activated) end up with the new default `linkDestination` applied, but the `href` isn't set.
- If such a plugin is disabled, the `linkDestination` reverts to a default of 'None', but image blocks added while the plugin was active still have an `href` value.

Not sure what the best way to solve those issues is. 🤔

jesusangel commented:
Hi,

I've made a pull request (#14398) that fixes @talldan comment:

Also in testing I noticed a few things about the general approach of using a plugin to change attribute defaults in this way:

- Existing image blocks (added before the plugin was activated) end up with the new default linkDestination applied, but the href isn't set.
- If such a plugin is disabled, the linkDestination reverts to a default of 'None', but image blocks added while the plugin was active still have an href value.

I've added some code in ComponentDidMount method to figure out which is the right value for the LinkDestination attribute. For example, supose that you have changed the default value for LinkDestination from none to media and you edit some old post with some images in it. If the block markup don't have a value for the LinkDestination, then that's because the default value was "none", so the image isn't linked to any URL, but now the default value is "media". My code checks that the image hasn't got an anchor (<a href=""><img /></a>) and sets its LinkDestination value to none.

Hope this helps.

Kind regards,

okainov commented:
Any update on this one? Really would like to see it in!

youknowriad commented:
Hi There! I've triaging PRs today and it seems this one is stale. I'm going to close now but please consider reopening if you have some time to dedicate to it. Thanks for your efforts.

SeanDS commented:
@youknowriad I think this was invalidated by the changes to the image block anyway, namely how the URL is picked via a new popup instead of the sidebar. I'm still interested in getting this to work though, as it's annoying that the default URL target is still not possible to set.

thanh25896 commented:
When I added your code to function.php I got error : "Parse error: syntax error, unexpected ',', expecting variable (T_VARIABLE) "

I used lastest gutenberg plugin
